### DISCUSSION TITLE
finding information about users quicklisp and local-projects

### DISCUSSION THREAD

**Discussion by bigos** [kaveh808/kons-9#161] [Show and tell]
https://github.com/bigos/quicklisp-doctor/blob/bf75da1b70f1867fcc9825dfbcc3a4abe73e4e31/quicklisp-doctor.lisp#L43

Could the following be useful? 

```
QUICKLISP-DOCTOR> (print-relevant-info)
OS *************************
OS :LINUX
architecture :X64
quicklisp *************************
home #P"/home/jacek/quicklisp/"
client version "2021-02-13"
; Fetching #<URL "http://beta.quicklisp.org/dist/quicklisp-versions.txt">
; 9.13KB
==================================================
9,348 bytes in 0.00 seconds (0.00KB/sec)
latest dist version "2022-07-08"
loaded dist version "2022-07-08"
local projects (#P"/home/jacek/quicklisp/local-projects/")
git *************************
git path (0 "/usr/bin/git")
git version (0 "git version 2.34.1")
git commits in local folders *************************
/home/jacek/quicklisp/local-projects/cl-confidence/ GIT commit 14dc40a0fde822f4aab4fefc9b40ef5fdcbcd666
/home/jacek/quicklisp/local-projects/dot-cons-tree/ GIT commit 529dc869e6e46535f4d5567658d750068e55bd5b
/home/jacek/quicklisp/local-projects/graph-viewer/ NO-GIT-DETECTED 
/home/jacek/quicklisp/local-projects/hashtree/ GIT commit ca0eeb9a0d0bbe6d6dca8335f74f89b42bd1e6eb
/home/jacek/quicklisp/local-projects/tris/ NO-GIT-DETECTED 
```

**Comment by foretspaisibles** [comment]
I believe this is interesting for diagnostics, how do you suggest to use that?

Part of the information is covered by the `development/describe` script, as

```
✗ development/describe 
Uname: Darwin
System: macOS 13.0
Architecture: arm64
Date: Sun Oct 30 09:11:15 CET 2022
SBCL: /opt/local/bin/sbcl (SBCL 2.2.9)
Brew: /opt/homebrew/bin/brew
MacPorts: /opt/local/bin/port
QuickLisp: T
```

I believe the shell script entry point is nice because it is actionable in one step from the console. The describe script could be better called `describe-workstation` to something like that, yes, you get the idea.

How would you picture yourself a merge of the two ideas? Also it's great to have version information, maybe we could try to put the time of the last commit too, this helps users to see quickly if they are using the latest software or not.

If you are using Lisp to write this kind of scripts, did you consider Rashell? (Shameless self-advertisement here :-)  https://github.com/melusina-org/cl-rashell/. )

**Reply by bigos** [reply]
It is a demonstration of using SBCL to find information about the environment and can be a starting point for something we can decide to try at a later stage.

You are correct about easier scenarios. My solution is waiting for a more challenging problem.

I have added esrap to the asd file. So I imagine parsing the output of the commands to get the details you suggest. And make different choices based on that information.

 Is it possible that the two ideas could diverge? But still, influence each other in a good way like your last suggestion?

**Reply by bigos** [reply]
I have added commit and date. 
https://github.com/bigos/quicklisp-doctor/blob/dbd85a67b60776605ca55482f9fefa1d9e52fca2/quicklisp-doctor.lisp#L31

**Reply by bigos** [reply]
Your cl-rashell readme page links to documentation do not work. It would be interesting to see how you want to solve some interesting problems. I can have a look later.

**Reply by bigos** [reply]
I have made my idea run on Windows. It is kind of ugly.

**Reply by bigos** [reply]
I have tidied up the code a bit and resolved some problems with the Windows machine. Now I need a bit of a break from it. I will come back to it and have another look at your suggestions. I do not have a Mac, so adding Mac-specific stuff could be a problem. What I have so far could easily be adapted to be run using sbcl --load. That would be a step towards merging the ideas. But the discussion elsewhere makes possible another approach. I do not know how this thing will unfold. Please suggest anything you feel would add to the discussion.

**Comment by bigos** [comment]
I have renamed the function as suggested.

```
QUICKLISP-DOCTOR> (describe-workstation)
OS *************************
OS :WIN
architecture :X64
quicklisp *************************
home #P"C:/msys64/home/jacek/quicklisp/"
client version "2021-02-13"
; Fetching #<URL "http://beta.quicklisp.org/dist/quicklisp-versions.txt">
; 9.13KB
==================================================
9,348 bytes in 0.00 seconds (8306.56KB/sec)
latest dist version "2022-07-08"
loaded dist version "2022-02-20"

!!!!!!!!!!!!!!!! WARNING your quicklisp is not up to date !!!!!!!!!!!!!!!!!!!!!!!!
In your Lisp REPL please run 
(ql:update-client)
(ql:update-dist)

local projects (#P"C:/msys64/home/jacek/quicklisp/local-projects/")
path *************************
depending onyour system Lisp and editor amy have different paths
current path C:\msys64\mingw64\bin;C:\msys64\usr\local\bin;C:\msys64\usr\bin;C:\msys64\usr\bin;C:\Windows\System32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\msys64\usr\bin\site_perl;C:\msys64\usr\bin\vendor_perl;C:\msys64\usr\bin\core_perl;C:\Program Files\Emacs\x86_64\bin;C:\Program Files\Steel Bank Common Lispgit *************************
git path (NIL "/usr/bin/git")
WARNING: using git-path "c:/msys64/usr/bin/git.exe"
git version (NIL "git version 2.35.1")
git commits in local folders *************************
C:/msys64/home/jacek/quicklisp/local-projects/quicklisp-doctor/ GIT (commit c790400181503482b8631dcda8b2e1b2ea5a4351
                                                                     Date:   Sun Oct 30 21:20:20 2022 +0000)

```

**Reply by bigos** [reply]
https://github.com/bigos/quicklisp-doctor/blob/88609f6ea807dad9a5b7720f98f52a21f23e4a3d/quicklisp-doctor.lisp#L126

I had in mind something like this. Should I go down this path again? I guess we are switching to git-submodules, but still, it was an interesting exercise.
